File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/g_longest_consecutive_sequence.rs
use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let mut max_len = 0;
        let mut from_to = HashMap::new();
        let mut to_from = HashMap::new();
        for num in nums {
            if from_to.contains_key(&num) || to_from.contains_key(&num) {
                continue;
            }
            let mut right: Option<i32> = None;
            let mut left: Option<i32> = None;
            from_to.entry(num + 1).and_modify(|v| { right = Some(*v) });
            to_from.entry(num - 1).and_modify(|v| { left = Some(*v) });
            match right {
                Some(r) => {
                    match left {
                        Some(l) => {
                            from_to.remove(&(num + 1));
                            from_to.insert(l, r);
                            to_from.remove(&(num - 1));
                            to_from.insert(r, l);
                            max_len = max_len.max(r - l + 1);
                        }
                        None => {
                            from_to.remove(&(num + 1));
                            from_to.insert(num, r);
                            to_from.insert(r, num);
                            max_len = max_len.max(r - num + 1);
                        }
                    }
                }
                None => {
                    match left {
                        Some(l) => {
                            from_to.insert(l, num);
                            to_from.remove(&(num - 1));
                            to_from.insert(num, l);
                            max_len = max_len.max(num - l + 1);
                        }
                        None => {
                            from_to.insert(num, num);
                            to_from.insert(num, num);
                            max_len = max_len.max(1);
                        }
                    }
                }
            }
        }
        return max_len;
    }
}

pub fn test() {
    println!("longest_consecutive_sequence");
    println!("{}", 4 == Solution::longest_consecutive(vec![100, 4, 200, 1, 3, 2]));
    println!("{}", 9 == Solution::longest_consecutive(vec![0, 3, 7, 2, 5, 8, 4, 6, 0, 1]));
    println!("{}", 4 == Solution::longest_consecutive(vec![-7, -1, 3, -9, -4, 7, -3, 2, 4, 9, 4, -9, 8, -7, 5, -1, -7]));
}

File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/g_longest_consecutive_sequence.ts
function longestConsecutive(nums: number[]): number {
    let max_len = 0
    const from_to = new Map<number, number>()
    const to_from = new Map<number, number>()
    for (const num of nums) {
        if (from_to.has(num) || to_from.has(num)) {
            continue
        }
        const right = from_to.get(num + 1)
        const left = to_from.get(num - 1)
        if (right != undefined) {
            if (left != undefined) {
                from_to.delete(num + 1)
                from_to.set(left, right)
                to_from.delete(num - 1)
                to_from.set(right, left)
                max_len = Math.max(max_len, right - left + 1)
            } else {
                from_to.delete(num + 1)
                from_to.set(num, right)
                to_from.set(right, num)
                max_len = Math.max(max_len, right - num + 1)
            }
        } else {
            if (left != undefined) {
                from_to.set(left, num)
                to_from.delete(num - 1)
                to_from.set(num, left)
                max_len = Math.max(max_len, num - left + 1)
            } else {
                from_to.set(num, num)
                to_from.set(num, num)
                max_len = Math.max(max_len, 1)
            }
        }
    }
    return max_len
}

console.log(4 == longestConsecutive([100, 4, 200, 1, 3, 2]))
console.log(9 == longestConsecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1]))
console.log(4 == longestConsecutive([-7, -1, 3, -9, -4, 7, -3, 2, 4, 9, 4, -9, 8, -7, 5, -1, -7]))
